### DISCUSSION TITLE
Use static method to build immutable type (`org.apache.commons.lang3.Range<Long>`)

### DISCUSSION THREAD

**Discussion by luca-bassoricci** [mapstruct/mapstruct#3181] [Q&A]
```
class RangeDTO {
  public org.apache.commons.lang3.Range<Long> range;
}

class SourceDTO {
  public Long min;
  public Long max;
}

public interface Mapper {
	@Mapping(target = "range.max", source = "max")
	@Mapping(target = "range.min", source = "min")
	public abstract RangeDTO map(SourceDTO source);
}
```

This mapper method produces `Range<Long> does not have an accessible constructor` error (it's fine, `Range` doesn't have an accessible constructor) because it must created using static method [Range#between()](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Range.html#between-T-T-).

I prefer not to use `expression` nor writing a `Range<Long> map(SourceDTO)`.
Any advice?
Thanks in advance
Luca

**Comment by asjp1970** [comment]
Hi Luca,

I'd suggest you start with [Custom builder provider](https://mapstruct.org/documentation/stable/reference/html/#custom-builder-provider). At the end of the day, a builder is what you use case needs, because Range doesn't have public constructors.
The only problem I see is that I've never used a builder in MapStruct requiring the source parameters as input, so I'm not sure how I would do that... But it seems like you need a customized building strategy

BR/Alberto

**Reply by filiphr** [reply]
This is a really good proposal @asjp1970. The best solution for this would be to implement your own builder for `org.apache.commons.lang3.Range` and then instruct MapStruct to use that one.
